metals/src/main/scala/scala/meta/internal/metals/onTypeRangeFormatters/IndentOnPaste.scala

```diff
@@ -0,0 +1,173 @@
+  val indentRegex: Regex = raw"\S".r
+
+  private def codeStartPosition(line: String): Option[Int] =
+    indentRegex.findFirstMatchIn(line).map(_.start)
+
+  private def increaseIndentation(line: String) =
+    increaseIndentPatternRegex.findFirstIn(line).nonEmpty
+
+  private def stringRepeat(s: String, n: Int): String =
+    ("%0" + n + "d").format(0).replace("0", s)
+
+  // converts spaces into tabs and vice-versa, normalizing the lengths of indentations
+  private def normalizeSpacesAndTabs(
+      line: String,
+      codeStartPos: Int,
+      blank: String,
+      tabSize: Int
+  ): String = {
+    if (codeStartPos != 0) {
+      val substrLength = math.min(line.length, codeStartPos)
+      val indentation = line.substring(0, substrLength)
+      val indentChars = indentation.split("")
+      val pastedBlank = indentChars.head
+      blank match {
+        case "\t" if pastedBlank == blank => line
+        case " " if pastedBlank == blank => line
+        case "\t" if pastedBlank == " " =>
+          val tabNum = math.ceil(pastedBlank.length / 2).toInt
+          stringRepeat(blank, tabNum) ++ line.slice(codeStartPos, line.length)
+        case " " if pastedBlank == "\t" =>
+          stringRepeat(blank, tabSize) ++ line.slice(
+            codeStartPos,
+            line.length
+          )
+        case _ => line
+      }
+    } else line
+  }
+
+  override def contribute(
+      sourceText: String,
+      range: lsp4j.Range,
+      splitLines: Array[String],
+      startPos: StartPosition,
+      endPos: EndPosition,
+      formattingOptions: FormattingOptions,
+      tokens: Option[Tokens]
+  ): Option[List[TextEdit]] = {
+    val insertSpaces = formattingOptions.isInsertSpaces
+    val originalTabSize = formattingOptions.getTabSize
+    val rangeStart = startPos.toLSP.getStart
+    rangeStart.setCharacter(0)
+    val pastedRange = new Range(rangeStart, endPos.toLSP.getEnd)
+    val startLine = startPos.toLSP.getStart.getLine
+    val endLine = endPos.toLSP.getEnd.getLine
+
+    val splitLinesWithIndex = splitLines.zipWithIndex
+    val inRangeLines = splitLinesWithIndex.slice(startLine, endLine + 1)
+    val pastedLines = inRangeLines.map(_._1)
+    val pastedLinesWithIndex = pastedLines.zipWithIndex
+
+    val (blank, tabSize) =
+      if (insertSpaces) (" ", originalTabSize) else ("\t", 1)
+
+    // These are the lines from the first pasted line, going above
+    val prePastedLines = splitLinesWithIndex.take(startLine).reverse
+
+    val currentIndentationLevel = (for {
+      (line, _) <- prePastedLines.find(t => {
+        val trimmed = t._1.trim()
+        trimmed.nonEmpty && !trimmed.startsWith("|")
+      }) // Find first line non empty (aka code) that is not a piped multi-string
+
+      indentation <- codeStartPosition(line) // get indentation spaces
+      nextIncrease = increaseIndentation(
```

Comment:
This part of the code checks if the first pasted line needs to have increase indentation compared to the non-pasted line.
Suppose you're pasting right after an if:
```scala
def function():
  if (1 < 2):
    @@ pasting here
```
Since we're not assuming anymore that we have the first line indented correctly due to @laughedelic suggestion,  we need to compute the first line correct indentation based on the previous one.
The rest of the code, essentially uses the idea you propose.

Also related to this, @dos65 tested that vim-mode for VSCode doesn't actually trigger `rangeFormatting` when pasting, but maybe using `coc-metals` it does?